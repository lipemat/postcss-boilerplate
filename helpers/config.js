const packageConfig = require( './package-config' );
const path = require( 'path' );


/**
 * Get a config for our /config directory merged with any
 * matching configuration from the project directory.
 *
 * For instance if we have a file named config/babel.config.js in our project
 * we will merge the contents with our config/babel.config.js in favor of whatever
 * is specified with the project's file.
 *
 * @param {String} fileName
 * @returns {object}
 */
function getConfig( fileName ) {
	let config = require( '../config/' + fileName );
	try {
		let localConfig = require( path.resolve( packageConfig.workingDirectory + '/config', fileName ) );
		config = {...config, ...localConfig};
	} catch ( e ) {
	}
	return config;
}

module.exports = {
	getConfig: getConfig
};
