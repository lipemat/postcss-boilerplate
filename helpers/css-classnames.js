const SHORT_ALPHABET = 'abcdefghijklmnopqrstuvwxyz';
const ALPHABET = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';

const classes = {};
let counters = [ -1 ];

/**
 * Reset all counters.
 *
 * @notice Mostly here for unit tests.
 */
function resetCounters() {
	counters = [ -1 ];
}

/**
 * Get the next class is sequence based on:
 * 1. Single character from SHORT_ALPHABET (prevent conflicts with JS boilerplate).
 * 2. Incremented character from the `ALPHABET`.
 *      1. Used once require 2+ characters.
 *      2. Grows to 3+ characters as needed.
 *
 * @return {string}
 */
function getNextClass() {
	const last = counters.length - 1;
	let totalLetters = ALPHABET.length - 1;

	// First level uses the SHORT_ALPHABET.
	if ( 0 === last ) {
		totalLetters = SHORT_ALPHABET.length - 1;
	}

	if ( counters[ last ] < totalLetters ) {
		counters[ last ]++;
	} else {
		incrementParent();
	}

	return counters.map( ( counter, i ) => {
		return 0 === i ? SHORT_ALPHABET[ counter ] : ALPHABET[ counter ];
	} ).join( '' );
}


/**
 * When we run out of characters on the current level:
 * 1. Increment the parent level.
 * 2. Reset current level and all child levels back to 0.
 *
 * If we are out of characters on the parent level or have
 * no parent level:
 * 1. Add a new child level.
 * 2. Reset all levels back to 0.
 *
 */
function incrementParent() {
	let parent = counters.length - 2;
	let totalLetters = ALPHABET.length - 1;

	while ( counters[ parent ] !== undefined ) {
		// First level uses the SHORT_ALPHABET.
		if ( 0 === parent ) {
			totalLetters = SHORT_ALPHABET.length - 1;
		}
		if ( counters[ parent ] < totalLetters ) {
			counters[ parent ]++;
			// Reset all child levels to 0.
			while ( counters[ parent + 1 ] !== undefined ) {
				counters[ parent + 1 ] = 0;
				parent++;
			}
			return;
		}
		parent--;
	}

	// Add a new level and reset all existing levels.
	counters.forEach( ( _, i ) => counters[ i ] = 0 );
	counters.push( 0 );
}

/**
 * Return a single character unique CSS class name based on
 * postcss-modules's `generateScopedName` callback.
 *
 * Tracks CSS classes per each file so duplicate uses of the
 * same class in a file receive the same result.
 *
 * @notice Only enabled if the `package.json` has `shortCssClasses` set to true.
 *
 * @link https://github.com/madyankin/postcss-modules#generating-scoped-names
 */
const generateScopedName = ( localName, resourcePath ) => {
	classes[ resourcePath ] ||= {};
	classes[ resourcePath ][ localName ] ||= getNextClass();
	return classes[ resourcePath ][ localName ];
};

module.exports = {
	ALPHABET,
	SHORT_ALPHABET,
	generateScopedName,
	getNextClass,
	resetCounters,
};
