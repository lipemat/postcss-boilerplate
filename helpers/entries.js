const config = require( './package-config' );
const fs = require( 'fs' );
const path = require( 'path' );
const {getConfig} = require( './config' );

const entries = getConfig( 'postcss-entries.js' );


/**
 * Entry points to be loaded by Grunt.
 *
 * Checks each provided file for exists and includes if it does.
 *
 * @see config/postcss-entries.js
 */
function getEntries() {
	const matches = {
		min: {},
		toCSS: {}
	};

	Object.values( entries ).forEach( possibleFile => {
		const filePath = config.theme_path + 'pcss/' + possibleFile;
		if ( fs.existsSync( path.resolve( filePath + '.pcss' ) ) ) {
			matches.toCSS[ config.css_folder + possibleFile + '.css' ] = config.theme_path + `pcss/${possibleFile}.pcss`;
			matches.min[ config.css_folder + possibleFile + '.min.css' ] = config.theme_path + `pcss/${possibleFile}.pcss`;
		}
	} );
	return matches;
}

module.exports = {
	getEntries,
};
