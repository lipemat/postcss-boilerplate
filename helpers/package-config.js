/**
 * Get all configurations for package.json of the project running this.
 */
const path = require( 'path' );
const fs = require( 'fs' );
const workingDirectory = fs.realpathSync( process.cwd() );

let packageConfig = require( path.resolve( workingDirectory, 'package.json' ) );
packageConfig.workingDirectory = workingDirectory;
packageConfig.theme_path = packageConfig.theme_path || '';

try {
	let localConfig = require( path.resolve( workingDirectory, './local-config.json' ) );
	packageConfig = { ...packageConfig, ...localConfig }
} catch ( e ) {
}

module.exports = packageConfig;
