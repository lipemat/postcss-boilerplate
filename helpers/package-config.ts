import path from 'path';
import fs from 'fs';

/**
 * Get all configurations for package.json of the project running this.
 */
const workingDirectory = fs.realpathSync( process.cwd() );

let packageConfig = require( path.resolve( workingDirectory, 'package.json' ) );
packageConfig.brotliFiles ||= false;
packageConfig.workingDirectory = workingDirectory;
packageConfig.theme_path ||= './';
// Could be set to "" which would always test false.
if ( ! Boolean( packageConfig.hasOwnProperty( 'css_folder' ) ) ) {
	packageConfig.css_folder = 'css/';
}
packageConfig.combinedJson ||= false;
packageConfig.file_name ||= 'front-end';
packageConfig.shortCssClasses ||= false;

try {
	const localConfig = require( path.resolve( workingDirectory, './local-config.json' ) );
	packageConfig = {...packageConfig, ...localConfig};
} catch ( e ) {
}

/**
 * Helper function to get the results of `packageConfig`.
 *
 * - Allows mocking the results of `packageConfig` for testing.
 * - Allows getting the config through a callback instead of an import.
 *
 * @since 4.6.0
 */
export function getPackageConfig() {
	return packageConfig;
}

packageConfig.getPackageConfig = getPackageConfig;

// Leaving old export structure for backwards compatibility.
// @todo Remove in favor of default export in version 5.
module.exports = packageConfig;
