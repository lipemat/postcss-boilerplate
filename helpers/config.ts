import fs from 'fs';
import path from 'path';
import browserslist from 'browserslist';

import {getPackageConfig} from './package-config';


/**
 * Get a config for our /config directory merged with any
 * matching configuration from the project directory.
 *
 * For instance if we have a file named config/babel.config.js in our project
 * we will merge the contents with our config/babel.config.js in favor of whatever
 * is specified with the project's file.
 */
export function getConfig( fileName: string ) {
	let config = require( '../config/' + fileName );
	const packageConfig = getPackageConfig();
	try {
		const localConfig = require( path.resolve( packageConfig.workingDirectory + '/config', fileName ) );
		if ( 'function' === typeof localConfig ) {
			config = {...config, ...localConfig( config )};
		} else {
			config = {...config, ...localConfig};
		}
	} catch ( e ) {
	}
	return config;
}

/**
 * Get the browserslist from the current project.
 *
 * - If specified using standard browserslist config, we will use that.
 * - Fallback to WordPress defaults except for "> 1%".
 */
export function getBrowsersList() {
	const projectBrowsersList = browserslist();
	if ( browserslist( browserslist.defaults ) === projectBrowsersList ) {
		const wp = [ ...require( '@wordpress/browserslist-config' ) ];
		wp.push( 'not and_uc 15.5' );
		return wp;
	}
	return projectBrowsersList;
}

/**
 * If browserslist is not specified, we fall back to WordPress defaults.
 *
 * Return false if a browserslist is specified in the current project.
 *
 * @deprecated Use getBrowsersList() instead.
 *
 * @link https://github.com/browserslist/browserslist#config-file
 *
 * @return {false | string[]}
 */
export const getDefaultBrowsersList = () => {
	if ( browserslist( browserslist.defaults ) === browserslist() ) {
		const wp = [ ...require( '@wordpress/browserslist-config' ) ];
		wp.push( 'not and_uc 15.5' );
		return wp;
	}
	return false;
};


/**
 * Provide CSS properties and media queries to all postcss plugins.
 *
 * If a media-queries files exist, automatically load them.
 * If CSS variables exist, automatically load them.
 *
 * 1. pcss/globals/variables.pcss
 * 3. pcss/globals/media-queries.pcss
 */
export function getExternalFiles(): string[] {
	const externalFiles: string[] = [];
	const packageConfig = getPackageConfig();

	[
		path.resolve( packageConfig.theme_path, 'pcss/globals/media-queries.pcss' ),
		path.resolve( packageConfig.theme_path, 'pcss/globals/variables.pcss' ),
	].forEach( possibleFile => {
		if ( fs.existsSync( possibleFile ) ) {
			externalFiles.push( possibleFile );
		}
	} );

	return externalFiles;
}
