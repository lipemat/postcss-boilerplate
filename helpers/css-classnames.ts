import {getPackageConfig} from './package-config';

export const SHORT_ALPHABET = 'abcdefghijklmnopqrstuvwxyz';
export const ALPHABET = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';

const classes: {
	[ filename: string ]: {
		[ className: string ]: string
	}
} = {};

let counters = [ -1 ];

/**
 * Check if short CSS classes are enabled.
 *
 * Using a helper function to allow for future enhancements.
 *
 * @since 4.6.0
 */
export function usingShortCssClasses(): boolean {
	return Boolean( getPackageConfig().shortCssClasses );
}

/**
 * Reset all counters.
 *
 * @notice Mostly here for unit tests.
 */
export function resetCounters() {
	counters = [ -1 ];
}

/**
 * Get the next class is sequence based on:
 * 1. Single character from SHORT_ALPHABET (prevent conflicts with JS boilerplate).
 * 2. Incremented character from the `ALPHABET`.
 *      1. Used once require 2+ characters.
 *      2. Grows to 3+ characters as needed.
 *
 * @return {string}
 */
export function getNextClass() {
	const last = counters.length - 1;
	let totalLetters = ALPHABET.length - 1;

	// First level uses the SHORT_ALPHABET.
	if ( 0 === last ) {
		totalLetters = SHORT_ALPHABET.length - 1;
	}

	if ( counters[ last ] < totalLetters ) {
		counters[ last ]++;
	} else {
		incrementParent();
	}

	return counters.map( ( counter, i ) => {
		return 0 === i ? SHORT_ALPHABET[ counter ] : ALPHABET[ counter ];
	} ).join( '' );
}


/**
 * When we run out of characters on the current level:
 * 1. Increment the parent level.
 * 2. Reset current level and all child levels back to 0.
 *
 * If we are out of characters on the parent level or have
 * no parent level:
 * 1. Add a new child level.
 * 2. Reset all levels back to 0.
 *
 */
function incrementParent() {
	let parent = counters.length - 2;
	let totalLetters = ALPHABET.length - 1;

	while ( counters[ parent ] !== undefined ) {
		// First level uses the SHORT_ALPHABET.
		if ( 0 === parent ) {
			totalLetters = SHORT_ALPHABET.length - 1;
		}
		if ( counters[ parent ] < totalLetters ) {
			counters[ parent ]++;
			// Reset all child levels to 0.
			while ( counters[ parent + 1 ] !== undefined ) {
				counters[ parent + 1 ] = 0;
				parent++;
			}
			return;
		}
		parent--;
	}

	// Add a new level and reset all existing levels.
	counters.forEach( ( _, i ) => counters[ i ] = 0 );
	counters.push( 0 );
}

/**
 * Return a single character unique CSS class name based on
 * postcss-modules's `generateScopedName` callback.
 *
 * Tracks CSS classes per each file so duplicate uses of the
 * same class in a file receive the same result.
 *
 * @notice Only enabled if the `package.json` has `shortCssClasses` set to true.
 *
 * @link https://github.com/madyankin/postcss-modules#generating-scoped-names
 */
export const generateScopedName = ( localName: string, resourcePath: string ): string => {
	classes[ resourcePath ] ||= {};
	classes[ resourcePath ][ localName ] ||= getNextClass();
	return classes[ resourcePath ][ localName ];
};


/**
 * Get the hash to generate the CSS module name, or
 * the `generateScopeName` for short CSS classes
 * if enabled.
 *
 * @note If run into issues with class name conflicts @see b36fc5309 as a more robust alternative.
 */
export function getGenerateScopeName() {
	if ( 'production' === process.env.NODE_ENV ) {
		// Use short CSS classes if enabled.
		if ( usingShortCssClasses() ) {
			return generateScopedName;
		}
		return '[contenthash:base52:5]';
	}
	return 'â“‚[name]__[local]__[contenthash:base52:2]';
}
