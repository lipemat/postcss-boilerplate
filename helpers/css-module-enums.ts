import type {Environment} from './config';
import path from 'path';
import {getPackageConfig} from './package-config';
import ejs from 'ejs';
import fse from 'fs-extra';

/**
 * Get the top level folder from where the CSS is compiled.
 *
 * @example './css-root/dist/' => config.theme_path + 'css-root'
 */
export function getDistFolder() {
	const config = getPackageConfig();
	const directories = config.css_folder
		.split( '/' )
		.filter( ( part: string ) => '' !== part && '.' !== part );

	return path
		.resolve( config.theme_path + ( directories[ 0 ] ?? '' ) )
		.replace( /\\/g, '/' );
}


function getCombinedName( env: Environment ): string {
	return 'production' === env ? 'module-enums.min.php' : 'module-enums.php';
}


function convertPathToEnum( fullPath: string ): string {
	const pathWithoutExtension = fullPath.replace( '.pcss/', '' );
	const parts = pathWithoutExtension.split( '/' );
	return parts.map( word => {
		return word.split( '-' ).map( subWord => {
			return subWord.charAt( 0 ).toUpperCase() + subWord.slice( 1 );
		} ).join( '_' );
	} ).join( '__' );
}


function convertToCamelCase( cssClass: string ): string {
	const words = cssClass.split( '-' );
	for ( let i = 1; i < words.length; i++ ) {
		words[ i ] = words[ i ].charAt( 0 ).toUpperCase() + words[ i ].slice( 1 );
	}
	return words.join( '' );
}


/**
 * Create a PHP file with an enum for each CSS module.
 *
 * Uses `ejs` to render the template.
 */
export class CssModuleEnums {
	private readonly json: Object;
	private readonly filePath: string;


	constructor( filePath: string, json: Object ) {
		this.filePath = filePath;
		this.json = json;
	}


	/**
	 * Add the module to the combined PHP enum file.
	 */
	addModuleToEnum( env: Environment ) {
		const combined = ( getDistFolder() + '/' + getCombinedName( env ) ).replace( /\\/g, '/' );
		let content = '';
		try {
			content = fse.readFileSync( combined, 'utf-8' ) ?? '';
		} catch ( e ) {
		}
		if ( '' === content ) {
			content = fse.readFileSync( __dirname + '/templates/module-enum-header.ejs', 'utf-8' );
		}

		const template = fse.readFileSync( __dirname + '/templates/module-enum.ejs', 'utf-8' );
		content += ejs.render( template, {
			classMap: this.getFormattedClassMap(),
			className: convertPathToEnum( this.filePath ),
		} );
		fse.outputFileSync( combined, content );
	}


	/**
	 * Get the JSON classmap with all keys converted to camelCase.
	 */
	private getFormattedClassMap(): Object {
		return Object.keys( this.json )
			.map( key => convertToCamelCase( key ) )
			.reduce( ( result, key, index ) => {
				result[ key ] = Object.values( this.json )[ index ];
				return result;
			}, {} );
	}
}
