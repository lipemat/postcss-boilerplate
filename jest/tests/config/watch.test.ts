import {getPackageConfig} from '../../../helpers/package-config';

let mockWatch: false | string[] = false;
let mockCssEnums: boolean = false;

// Change the result of the getPackageConfig function.
jest.mock( '../../../helpers/package-config.ts', () => ( {
	...jest.requireActual( '../../../helpers/package-config.ts' ),
	getPackageConfig: () => {
		const pkgConfig = jest.requireActual( '../../../helpers/package-config.ts' );
		pkgConfig.cssEnums = mockCssEnums;
		if ( false !== mockWatch ) {
			pkgConfig.pcssWatch = mockWatch;
		}
		return pkgConfig;
	},
} ) );

function getWatchConfig() {
	jest.resetModules();
	return require( '../../../config/watch.ts' );
}

afterEach( () => {
	mockWatch = false;
} );

describe( 'Test watch config', () => {
	it( 'watch files from package.json', () => {
		expect( getWatchConfig().postcss.files ).toStrictEqual( [
			getPackageConfig().theme_path + 'pcss/**/*.{pcss,css}',
			getPackageConfig().theme_path + 'template-parts/**/*.{pcss,css}',
		] );

		mockWatch = [ 'woocommerce', 'template-parts', 'pcss' ];
		expect( getWatchConfig().postcss.files ).toStrictEqual( [
			getPackageConfig().theme_path + 'woocommerce/**/*.{pcss,css}',
			getPackageConfig().theme_path + 'template-parts/**/*.{pcss,css}',
			getPackageConfig().theme_path + 'pcss/**/*.{pcss,css}',
		] );
	} );

	it( 'matches snapshot', () => {
		mockCssEnums = true;
		expect( getWatchConfig() ).toMatchSnapshot();

		mockWatch = [ 'woocommerce', 'template-parts', 'pcss' ];
		expect( getWatchConfig() ).toMatchSnapshot();

		mockCssEnums = false;
		expect( getWatchConfig() ).toMatchSnapshot();
	} );
} );
