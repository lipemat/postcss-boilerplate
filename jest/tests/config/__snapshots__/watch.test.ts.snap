// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test watch config matches snapshot 1`] = `
{
  "php": {
    "files": [
      "./**/*.php",
      "!./css/**/*.php",
    ],
    "options": {
      "livereload": true,
    },
  },
  "postcss": {
    "files": [
      "./pcss/**/*.{pcss,css}",
      "./template-parts/**/*.{pcss,css}",
    ],
    "options": {
      "livereload": false,
      "spawn": false,
    },
    "tasks": [
      "caching:reset",
      "postcss:toCSS",
      "caching:writeModules:development",
      "caching:reload",
    ],
  },
}
`;

exports[`Test watch config matches snapshot 2`] = `
{
  "php": {
    "files": [
      "./**/*.php",
      "!./css/**/*.php",
    ],
    "options": {
      "livereload": true,
    },
  },
  "postcss": {
    "files": [
      "./woocommerce/**/*.{pcss,css}",
      "./template-parts/**/*.{pcss,css}",
      "./pcss/**/*.{pcss,css}",
    ],
    "options": {
      "livereload": false,
      "spawn": false,
    },
    "tasks": [
      "caching:reset",
      "postcss:toCSS",
      "caching:writeModules:development",
      "caching:reload",
    ],
  },
}
`;

exports[`Test watch config matches snapshot 3`] = `
{
  "php": {
    "files": [
      "./**/*.php",
      "!./css/**/*.php",
    ],
    "options": {
      "livereload": true,
    },
  },
  "postcss": {
    "files": [
      "./woocommerce/**/*.{pcss,css}",
      "./template-parts/**/*.{pcss,css}",
      "./pcss/**/*.{pcss,css}",
    ],
    "options": {
      "livereload": false,
      "spawn": false,
    },
    "tasks": [
      "caching:reset",
      "postcss:toCSS",
      "caching:writeModules:development",
    ],
  },
}
`;
